generator/src/main/java/com/scottlogic/deg/generator/walker/reductive/ReductiveTreePruner.java

```diff
@@ -83,30 +77,61 @@ public ReductiveTreePruner(FieldSpecMerger merger, ConstraintReducer constraintR
 
-        if (!nodeFieldSpec.isPresent()) {
+        Merged<Map<Field, FieldSpec>> relevantFieldSpecs = createFieldSpecMap(relevantConstraints);
+        if (relevantFieldSpecs.isContradictory()){
             return Merged.contradictory();
         }
 
-        Optional<FieldSpec> newFieldSpec = merger.merge(nodeFieldSpec.get(), parentFieldSpec);
+        return mergeFieldSpecMaps(parentFieldSpecs, relevantFieldSpecs.get());
+    }
 
-        if (!newFieldSpec.isPresent()) {
-            return Merged.contradictory();
+    private Map<Field, Collection<AtomicConstraint>> getRelevantConstraints(Collection<AtomicConstraint> pulledUpAtomicConstraints, Set<Field> relevantFields) {
+        Map<Field, Collection<AtomicConstraint>> map = new HashMap<>();
+        for (AtomicConstraint constraint : pulledUpAtomicConstraints) {
+            if (relevantFields.contains(constraint.getField())) {
+                updateOrAddToList(map, constraint);
+            }
         }
+        return map;
+    }
 
-        return Merged.of(newFieldSpec.get());
+    private Merged<Map<Field, FieldSpec>> createFieldSpecMap(Map<Field, Collection<AtomicConstraint>> relevantConstraints){
+        Map<Field, FieldSpec> newFieldSpecs = new HashMap<>();
+        for (Map.Entry<Field, Collection<AtomicConstraint>> fieldToConstraints : relevantConstraints.entrySet()) {
+            Optional<FieldSpec> fieldSpec = constraintReducer.reduceConstraintsToFieldSpec(fieldToConstraints.getValue());
+            if (!fieldSpec.isPresent()){
+                return Merged.contradictory();
+            }
+            newFieldSpecs.put(fieldToConstraints.getKey(), fieldSpec.get());
+        }
+
+        return Merged.of(newFieldSpecs);
     }
 
-    private boolean onlyOneOption(Merged<DecisionNode> prunedDecisionNode) {
-        return prunedDecisionNode.get().getOptions().size() == 1;
+    private Merged<Map<Field, FieldSpec>> mergeFieldSpecMaps(Map<Field, FieldSpec> parentFieldSpecs, Map<Field, FieldSpec> newMap) {
+        for (Map.Entry<Field, FieldSpec> entry : parentFieldSpecs.entrySet()) {
+            if (!newMap.containsKey(entry.getKey())){
+                newMap.put(entry.getKey(), entry.getValue());
+            }else {
+                Optional<FieldSpec> mergedSpec = merger.merge(entry.getValue(), newMap.get(entry.getKey()));
+                if (!mergedSpec.isPresent()){
+                    return Merged.contradictory();
+                }
+                newMap.put(entry.getKey(), mergedSpec.get());
+            }
+        }
+
+        return Merged.of(newMap);
     }
 
-    private ConstraintNode getOnlyRemainingOption(Merged<DecisionNode> prunedDecisionNode) {
-        return prunedDecisionNode.get().getOptions().iterator().next();
+    private void updateOrAddToList(Map<Field, Collection<AtomicConstraint>> map, AtomicConstraint constraint) {
```

Comment:
This method always adds to a list, never updates?

It could be tidied up a bit to reduce nesting/else:

```
if (!map.containsKey(constraint.getField())) {
    map.put(constraint.getField(), new ArrayList<>());
}

map.get(constraint.getField()).add(constraint);
```